dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -0,0 +1,226 @@
+        Token gitlabToken = Token.extractToken(tokens, TokenType.GITLAB_COM);
+
+        final String bitbucketTokenContent = bitbucketToken == null ? null : bitbucketToken.getContent();
+        final String gitHubTokenContent = githubToken == null ? null : githubToken.getContent();
+        final String gitlabTokenContent = gitlabToken == null ? null : gitlabToken.getContent();
+
+        final SourceCodeRepoInterface sourceCodeRepo = SourceCodeRepoFactory
+            .createSourceCodeRepo(gitUrl, client, bitbucketTokenContent, gitlabTokenContent, gitHubTokenContent);
+        if (sourceCodeRepo == null) {
+            throw new CustomWebApplicationException("Git tokens invalid, please re-link your git accounts.", HttpStatus.SC_BAD_REQUEST);
+        }
+        return sourceCodeRepo;
+    }
+
+    /**
+     * Common code for upserting a version to a workflow/service
+     * @param repository Repository path (ex. dockstore/dockstore-ui2)
+     * @param gitReference Git reference (ex. master)
+     * @param user User calling endpoint
+     * @param workflowMode Mode of workflows to filter by
+     * @return Shared dockstore path to workflow/service
+     */
+    protected String upsertVersionHelper(String repository, String gitReference, User user, WorkflowMode workflowMode,
+            String installationAccessToken) {
+        // Create path on Dockstore (not unique across workflows)
+        String dockstoreWorkflowPath = String.join("/", TokenType.GITHUB_COM.toString(), repository);
+
+        // Find all workflows with the given path that are full
+        List<Workflow> workflows = findAllWorkflowsByPath(dockstoreWorkflowPath, workflowMode);
+
+        if (workflows.size() > 0) {
+            // All workflows with the same path have the same Git Url
+            String sharedGitUrl = workflows.get(0).getGitUrl();
+
+            // Set up source code interface and ensure token is set up
+            GitHubSourceCodeRepo sourceCodeRepo;
+            if (user != null) {
+                User updatedUser = userDAO.findById(user.getId());
+                sourceCodeRepo = (GitHubSourceCodeRepo)getSourceCodeRepoInterface(sharedGitUrl, updatedUser);
+            } else {
+                sourceCodeRepo = (GitHubSourceCodeRepo)SourceCodeRepoFactory.createGitHubAppRepo(installationAccessToken);
+            }
+
+            // Pull new version information from GitHub and update the versions
+            workflows = sourceCodeRepo.upsertVersionForWorkflows(repository, gitReference, workflows, workflowMode);
+
+            // Update each workflow with reference types
+            for (Workflow workflow : workflows) {
+                Set<WorkflowVersion> versions = workflow.getWorkflowVersions();
+                versions.forEach(version -> sourceCodeRepo.updateReferenceType(repository, version));
+            }
+        } else {
+            if (workflowMode == SERVICE) {
```

Comment:
This doesn't feel very abstract, maybe add a 
`//TODO` since it seems like this should apply to workflows when we get to them

Reply:
Added TODO.

For this PR, besides the findByPath fixes, I pretty much just moved things around and only changed things enough to get them working.